[FLINK-23139][state] Add TaskStateRegistry

rkhachatryan opened the pull request:
## What is the purpose of the change

Implement private state management for the ChangelogStateBackend.

[Detailed design doc](https://docs.google.com/document/d/1Nt95g5eWBkY5mFueF3dj639_4YHjhuGr_rn_6rmEjGw/edit?usp=sharing)
[Higher level design](https://docs.google.com/document/d/1NJJQ30P27BmUvD7oa4FChvkYxMEgjRPTVdO1dHLl_9I/edit#heading=h.9dxopqajsy7)

Completed parts:
- core classes (`package org.apache.flink.runtime.state.track;`)
- creation and injection into the `ChangelogKeyedStateBackend`
- most calls from the backend

Missing parts:
 - call `checkpointSubsumed` on subsumption (requires FLINK-21504)
 - ~call `stateUsed` on materialization (requires FLINK-21357)~ - FLINK-23344
 - call `stateUsed` on upload from DFS Changelog Writer (requires rebase as FLINK-21353 is now merged)
 - implement `getStateChangeUsageTracker` by DFS `StateChangelogStorage` (requires rebase as FLINK-21353 is now merged)

## Verifying this change

- Added `TaskStateRegistryImplTest` unit tests
- TBD: integration tests (after subsumption notification)

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

Myasuka commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/track/BackendStateRegistry.java:
```diff
@@ -0,0 +1,298 @@
+        }
+    }
+
+    private void snapshotSentForSavepoint(long checkpointId, Set<K> newStateKeys) {
```
From my point of view,  the name of `snapshotSentForSavepoint` is a bit hard to understand, `who` `send` the information to `where`?

rkhachatryan replied:
Actually the snapshot is created, not necessarily sent.
How about `savepointSnapshotCreated()`? (and `snapshotTaken` -> `snapshotCreated`)

Myasuka replied:
Do you think we could have `global savepoint`, `global checkpoint` and `local snapshot` terminologies to distinguish them?

rkhachatryan replied:
I think both checkpoints and checkpoints are currently global - tasks do not checkpoint independently of each other.
The snapshot can be global or local; but in TM, it's always local (unless everything fits into a single TM :)
So here it will be`local savepoint snapshot` or `local checkpoint snapshot` - and `local` seems redundant.
Or did I get you wrong?

Myasuka replied:
The task could have `local snapshot` successful after getting the state handle to return. And the `global checkpoint` might finally success or failed in the end, that's the duty of notification of checkpoint completed or aborted.
From this PR, I can see many place depends on the notifications, and if we could distinguish them, the code could look more clear.

Myasuka reviewed (commented):
Since this PR would be changed later. I mainly go through the logic but found it's still not easy to understand. Please take a look and answer my questions. Thanks.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/track/BackendStateRegistry.java:
```diff
@@ -0,0 +1,298 @@
+    private long lastSnapshottedCheckpoint = -1L;
+    @Nullable private Set<K> lastSnapshot = null;
+
+    private final Map<K, StateEntry<K>> inActiveUse = new HashMap<>();
```
I agree, will rename it.

rkhachatryan replied:
WDYT about renaming
`inActiveUse`
`inUseBySavepoints `
`inUseByCheckpoints `
to
`inUseByActiveCheckpoints` // in use by pending and potentially future savepoints AND checkpoints
`inUseByPastSavepoints`
`inUseByPastCheckpoints`
?

rkhachatryan replied:
I removed `inUseByPastSavepoints` as agreed offline and tentatively renamed 
`inActiveUse` -> `currentState`
`inUseBySavepoints` -> `previousStatesByHighestUsedCheckpoint`

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/track/TaskStateRegistry.java:
```diff
@@ -0,0 +1,243 @@
+     * @throws NoSuchElementException if the state is not registered
+     * @throws IllegalStateException if the given checkpoint was already performed
+     */
+    void snapshotTaken(
```
This is to make the difference explicit: that now a snapshot is created; the checkpoint was created (started) before and currently is in unknown state. `CompletedCheckpoint` might not be created yet (on JM); and TM might not even sent the ACK yet.
This difference is quite important in the logic of the component.

Myasuka replied:
I think you want to distinguish the checkpoint on local task side with the global checkpoint. If so, how about rename other methods starts with `snapshot`?

rkhachatryan replied:
Renamed to `stateSnapshotCreated` as agreed offline.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/track/TaskStateRegistryImpl.java:
```diff
@@ -0,0 +1,199 @@
+
+    @Override
+    public void close() throws Exception {
+        LOG.debug("Close");
```
Do you mean it should be removed or some info added?

Myasuka replied:
Yes, we could remove or add other information.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/track/StateEntry.java:
```diff
@@ -0,0 +1,259 @@
+    private static void putOnceIfNonEmpty(
+            Map<String, NavigableSet<Long>> target, NavigableSet<Long> set, String key) {
+        if (nonEmpty(set)) {
+            checkState(target.put(key, set) == null);
```
That would be more informative, I agree. Will change it.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/track/StateEntry.java:
```diff
@@ -0,0 +1,259 @@
+        this.key = key;
+        this.state = state;
+        this.cleaner = cleaner;
+        this.activelyUsingBackends = activelyUsingBackends;
```
Good point.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/track/TaskStateRegistry.java:
```diff
@@ -0,0 +1,243 @@
+ *       be deleted)
+ *   <li>for checkpoints, call either {@link #checkpointSubsumed(String, long) subsumed} or {@link
+ *       #checkpointAborted(String, long) aborted} (otherwise, state won't be deleted until {@link
+ *       #jobTerminated(JobStatus)})
```
The method was removed in 898fb0f1aa705f15d56e8a68f7a630c037b5af06, I'll update javadoc.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/track/TaskStateRegistry.java:
```diff
@@ -0,0 +1,243 @@
+     * Must be called <strong>after</strong> the corresponding {@link #checkpointStarting(String,
+     * long, boolean)}.
+     *
+     * @param trackStateUsage if true then new entries will be added and marked as used; any state
```
I think you are right about the usage, I'll remove the argument.
The differnece by `1L` was checked in the initial version and is not relevant anymore (I introduced `checkpointStarting` later which guarantees that a state won't be removed even if this `snapshotTaken` call is missing).
Sorry for the confusion.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:
```diff
@@ -208,6 +221,8 @@ public ChangelogKeyedStateBackend(
         this.mainMailboxExecutor = checkNotNull(mainMailboxExecutor);
         this.asyncOperationsThreadPool = checkNotNull(asyncOperationsThreadPool);
+        this.taskStateRegistry = checkNotNull(taskStateRegistry);
+        this.backendId = UUID.randomUUID().toString();
```
Good idea!

rkhachatryan commented:
Thanks for taking a look @Myasuka .
I've replied to your comments and updated the PR.
I agree that the logic is complex 🤷‍♂️. Please let me know if you have any questions (maybe we should have an offline discussion?)

rkhachatryan commented:
After the offline discussions I've changed the class structure and removed savepoints tracking. The algorithm is the same.
Do you mind @Myasuka if I squash the fixups? 
The PR also needs some rebasing to use changes from #15322

Myasuka commented:
I don't mind to squash all fixups, please go ahead.
